titus-server-master/src/main/java/com/netflix/titus/master/kubernetes/pod/DefaultPodAffinityFactory.java

```diff
@@ -247,11 +252,26 @@ private void processZoneConstraints() {
                 return;
             }
 
-            // If there is an EBS volume, it defaults to placement in the volume's zone
-            Optional<EbsVolume> optionalEbsVolume = EbsVolumeUtils.getEbsVolumeForTask(job, task);
-            if (optionalEbsVolume.isPresent()) {
-                addNodeAffinitySelectorConstraint(KubeConstants.NODE_LABEL_ZONE, optionalEbsVolume.get().getVolumeAvailabilityZone(), true);
-                return;
+            // If there is a Static IP allocation, get its zone
+            Optional<String> optionalIpAllocationZone = IpAddressAllocationUtils.getIpAllocationZoneForTask(job.getJobDescriptor(), task, titusRuntime.getCodeInvariants());
+
+            // If there is an EBS volume, get its zone
+            Optional<String> optionalEbsVolumeZone = EbsVolumeUtils.getEbsVolumeForTask(job, task)
+                    .map(EbsVolume::getVolumeAvailabilityZone);
+
+            if (optionalIpAllocationZone.isPresent() && optionalEbsVolumeZone.isPresent()) {
+                String ipAllocationZone = optionalIpAllocationZone.get();
+                String ebsVolumeZone = optionalEbsVolumeZone.get();
+                // If the zones for the two assigned resources do not match we do not assign either.
+                if (ebsVolumeZone.equals(ipAllocationZone)) {
```

Comment:
Since we are implicitly coupling the zones for both these attributes here, do you think it makes sense to update how we populate these taskContext attributes in the first place so there is a higher likelihood of zone match success between the two?